### PR TITLE
upstream sync 180607

### PR DESCRIPTION
**Pull request by whchung**
This is the PR to hold upstream sync on 180607.

### CODE REVIEW

**Comment by whchung** [comment]
@parallelo Let's use this upstream sync PR to once again test how CI works.

**Comment by parallelo** [comment]
Sounds good.  Btw - PR#16 dropped our build+test time from ~2hrs to down under 1 hr.  Will be looking at utilization to see if there are other tweaks we can make improvements on.

**Comment by whchung** [comment]
Looking at build error:
```
external/eigen_archive/unsupported/Eigen/CXX11/../../../Eigen/src/Core/util/Memory.h:168:37: error:  'Eigen::internal::handmade_aligned_malloc':  no overloaded function has restriction specifiers that are compatible with the ambient context 'aligned_malloc'
```
Strangely speaking `Eigen` used in `develop-upstream` branch is the same as the current tip, but we are only bumping into this issue here. Looking into it now.

**Comment by whchung** [comment]
@parallelo CI says:
```
Executed 471 out of 471 tests: 462 tests pass and [0m[31m[1m9 fail locally[0m.
```

Could you guide me how to siphon out failures from CI build logs?

**Comment by parallelo** [comment]
Here's the crude way of looking deeper.  
1. Go to the full log:  http://205.234.28.200:21096/job/tensorflow-upstream-unit-tests/83/consoleText
2. Find the `462 tests pass and 9 fail locally` metric at the bottom
3. Search for 'failed' and do some manual filtering

```
//tensorflow/python:function_def_to_graph_test                           [0m[31m[1mFAILED[0m in 105.3s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/function_def_to_graph_test/test.log
//tensorflow/python:gradient_descent_test                                [0m[31m[1mFAILED[0m in 92.0s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/gradient_descent_test/test.log
//tensorflow/python:momentum_test                                        [0m[31m[1mFAILED[0m in 133.1s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/momentum_test/test.log
//tensorflow/python/keras:models_test                                    [0m[31m[1mFAILED[0m in 82.1s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/models_test/test.log
//tensorflow/python/keras:training_test                                  [0m[31m[1mFAILED[0m in 96.4s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/training_test/test.log
//tensorflow/python/kernel_tests/distributions:beta_test                 [0m[31m[1mFAILED[0m in 50.4s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/kernel_tests/distributions/beta_test/test.log
//tensorflow/python/kernel_tests/distributions:dirichlet_test            [0m[31m[1mFAILED[0m in 116.2s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/kernel_tests/distributions/dirichlet_test/test.log
//tensorflow/python/kernel_tests/distributions:student_t_test            [0m[31m[1mFAILED[0m in 78.2s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/kernel_tests/distributions/student_t_test/test.log
//tensorflow/python/keras:cudnn_recurrent_test                           [0m[31m[1mFAILED[0m in 2 out of 2 in 114.8s
  Stats over 2 runs: max = 114.8s, min = 114.2s, avg = 114.5s, dev = 0.3s
  /var/lib/jenkins/workspace/tensorflow-upstream-unit-tests/bazel-ci_build-cache/.cache/bazel/_bazel_jenkins/eab0d61a99b6696edb3d2aff87b585e8/execroot/org_tensorflow/bazel-out/k8-opt/testlogs/tensorflow/python/keras/cudnn_recurrent_test/shard_2_of_2/test.log
```

Would be useful if we could create this summary without having to spend time digging.  I'm sure other folks have already done it -- will look around for improvements here.

**Comment by parallelo** [comment]
As a follow-up, to reproduce an individual issue using the same setup, you could run the following from the top of the `tensorflow-upstream` repo:  
```
tensorflow/tools/ci_build/ci_build.sh ROCM bazel test -c opt --config rocm //tensorflow/python/keras:training_test  
```
This command will start a docker container and kick off the test.

**Comment by whchung** [comment]
initial analysis:

| //tensorflow/python:function_def_to_graph_test |
| //tensorflow/python:gradient_descent_test |
| //tensorflow/python:momentum_test |
| //tensorflow/python/keras:models_test |
| //tensorflow/python/keras:training_test |
| //tensorflow/python/kernel_tests/distributions:beta_test |
| //tensorflow/python/kernel_tests/distributions:dirichlet_test |
| //tensorflow/python/kernel_tests/distributions:student_t_test |
| //tensorflow/python/keras:cudnn_recurrent_test |

**Comment by whchung** [comment]
@sunway513 per discussion with @dagamayank, will need your help to continue look into failing TensorFlow unit tests. I think reviewing failing tests introduced in this PR could be a good start.

**Comment by sunway513** [comment]
@whchung sounds good.

**Comment by whchung** [comment]
@parallelo I modified `ci_build` to mask 9 failed cases introduced in this PR:
https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/commit/bdc94ff54ceb72ae075d3537d69ab2f51194c30f

https://github.com/ROCmSoftwarePlatform/tensorflow-upstream/blob/develop-upstream-sync-180607/tensorflow/tools/ci_build/linux/rocm/run_py3_core.sh

But it seems the very latest CI build didn't pick it up:
http://205.234.28.200:21096/job/tensorflow-upstream-unit-tests/86/

http://205.234.28.200:21096/job/tensorflow-upstream-unit-tests/86/consoleFull

Could you help shed some lights on this? Would CI server use the `run_py3_core.sh` supplied in the PR, or it uses its own copy?

**Comment by whchung** [comment]
@parallelo moreover is it possible to force trigger a build for this CI?

**Comment by whchung** [comment]
retest this please

**Comment by parallelo** [comment]
retest this please

**Comment by parallelo** [comment]
retest this please

**Comment by parallelo** [comment]
retest this please
